azkv/keysource.go

```diff
@@ -0,0 +1,187 @@
+data key using Azure Key Vault with the Azure Go SDK.
+*/
+package azkv //import "go.mozilla.org/sops/azkv"
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"go.mozilla.org/sops/logging"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/2016-10-01/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/auth"
+	"github.com/sirupsen/logrus"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("AZKV")
+}
+
+// MasterKey is a Azure Key Vault key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	VaultURL string
+	Name     string
+	Version  string
+
+	EncryptedKey string
+	CreationDate time.Time
+}
+
+func newKeyVaultClient() (keyvault.BaseClient, error) {
+	var err error
+	c := keyvault.New()
+	c.Authorizer, err = auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		log.WithError(err).Error("Failed to create Azure authorizer")
+		return c, err
+	}
+
+	return c, nil
+}
+
+// NewMasterKey creates a new MasterKey from an URL, key name and version, setting the creation date to the current date
+func NewMasterKey(vaultURL string, keyName string, keyVersion string) *MasterKey {
+	return &MasterKey{
+		VaultURL:     vaultURL,
+		Name:         keyName,
+		Version:      keyVersion,
+		CreationDate: time.Now().UTC(),
+	}
+}
+
+// MasterKeysFromURLs takes a comma separated list of Azure Key Vault URLs and returns a slice of new MasterKeys for them
+func MasterKeysFromURLs(urls string) []*MasterKey {
+	var keys []*MasterKey
+	if urls == "" {
+		return keys
+	}
+	for _, s := range strings.Split(urls, ",") {
+		keys = append(keys, NewMasterKeyFromURL(s))
+	}
+	return keys
+}
+
+// NewMasterKeyFromResourceID takes an Azure Key Vault key URL and returns a new MasterKey
+// URL format is {vaultUrl}/keys/{key-name}/{key-version}
+func NewMasterKeyFromURL(url string) *MasterKey {
+	k := &MasterKey{}
+	re := regexp.MustCompile("^(https://[^/]+)/keys/([^/]+)/([^/]+)$")
+	parts := re.FindStringSubmatch(url)
+	if parts == nil || len(parts) < 2 {
+		log.Error("No match!")
+		// !?
+	}
+
+	k.VaultURL = parts[1]
+	k.Name = parts[2]
+	k.Version = parts[3]
+	k.CreationDate = time.Now().UTC()
+	return k
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with Key Vault and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	c, err := newKeyVaultClient()
+	if err != nil {
+		return err
+	}
+	data := base64.RawURLEncoding.EncodeToString(dataKey)
+	p := keyvault.KeyOperationsParameters{Value: &data, Algorithm: keyvault.RSAOAEP256}
```

Comment:
The choices of algorithm is RSA-OAEP, RSA-OAEP-256 and RSA-1_5. My understanding is that RSA-1_5 is broken, but I don't understand the deeper implications of the different hashing algorithms in use between OAEP and OAEP-256.

Reply:
I think all of those are considered secure, but I'd defer to @jvehent on that.

My main concern is that whatever we pick, the way this is implemented right now, it will be impossible to upgrade the algorithm in the future without breaking old files. Perhaps it would be worth it to default to RSA-OAEP-256, but store the used algorithm in the MasterKey, so that we use the same algorithm when decrypting as encrypting.

This isn't really a blocker to merge this, though. We could just make this change whenever we need to change the algorithm.